event-svc/src/event/validator/signed.rs

```diff
@@ -0,0 +1,267 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use ceramic_core::{SerializeExt, StreamId};
+use ceramic_event::unvalidated::{self, init, signed};
+use ceramic_validation::{cacao_verifier::Verifier, event_verifier::Verifier as _, VerifyJwsOpts};
+use cid::Cid;
+use ipld_core::ipld::Ipld;
+use tracing::warn;
+
+use crate::{
+    event::service::ValidationError,
+    store::{CeramicOneEvent, SqlitePool},
+    Result,
+};
+
+use super::{
+    event::{ValidatedEvent, ValidatedEvents},
+    grouped::{SignedEvents, Unsigned},
+    UnvalidatedEvent,
+};
+
+#[derive(Debug)]
+/// A signed validator is expected to be constructed to validate a batch of events and then be dropped.
+/// It will cache init events for the batch to use at different stages of validation and then drop everything
+/// when it finishes.
+pub struct SignedValidator {
+    /// The init events needed to validate this batch of events.
+    init_map: HashMap<Cid, Arc<unvalidated::Event<Ipld>>>,
+    /// The set of invalid init CIDs that can be used to reject any events that depend on them without even looking at them.
+    invalid_init: HashSet<Cid>,
+    /// The result of this validation batch that is updated as we go.
+    validated: ValidatedEvents,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum ValidationStatus {
+    ValidInit,
+    ValidData,
+    InvalidInit(String),
+    InvalidData(String),
+    Pending,
+}
+
+impl SignedValidator {
+    fn new(init_cnt: usize, data_cnt: usize) -> Self {
+        Self {
+            init_map: HashMap::with_capacity(init_cnt),
+            invalid_init: HashSet::new(),
+            validated: ValidatedEvents::new_with_expected_valid(init_cnt + data_cnt),
+        }
+    }
+
+    /// blindly trusts that this is indeed an init event
+    /// will error later if not
+    fn track_init(&mut self, cid: Cid, event: Arc<unvalidated::Event<Ipld>>) {
+        self.init_map.insert(cid, event);
+    }
+
+    fn add_invalid(&mut self, item: ValidationError) {
+        self.validated.invalid.push(item);
+    }
+
+    fn add_pending(&mut self, item: UnvalidatedEvent) {
+        self.validated.unvalidated.push(item);
+    }
+
+    fn add_valid(&mut self, item: ValidatedEvent) {
+        self.validated.valid.push(item);
+    }
+
+    fn add_valid_iter<I>(&mut self, item: I)
+    where
+        I: Iterator<Item = ValidatedEvent>,
+    {
+        self.validated.valid.extend(item);
+    }
+
+    pub async fn validate_events(
```

Comment:
Can we add a test for this method? It appears to be the main logic added in this PR.

Reply:
Yeah, probably worthwhile. I didn't add a new one because the existing tests all go through this logic (and started failing before updating the controller to a real one). I'll look at putting something dedicated in place.

Reply:
added some tests at the layer the service calls into (`EventValidator::validate_events`)

Reply:
there are tests related to different signature validation success/failure (did, cacao, invalid) in the Validation crate so I didn't focus on that here